File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/provider/model/NativeScaffoldModel.kt
package io.nativeblocks.runtime.api.provider.model

/**
 * Represents the scaffold structure in the native UI framework, consisting of a list of frame routes.
 */
data class NativeScaffoldModel(
    /**
     * List of frame routes included in the scaffold.
     */
    val frames: List<NativeFrameRouteModel>
)

/**
 * Represents a single frame route, including its metadata and route-specific details.
 */
data class NativeFrameRouteModel(
    /**
     * Unique identifier of the frame route.
     */
    val id: String? = null,

    /**
     * Name of the frame route.
     */
    val name: String? = null,

    /**
     * Type of the frame (e.g., FRAME, BOTTOM_SHEET, DIALOG).
     */
    val type: FrameTypeModel? = null,

    /**
     * Route path associated with the frame.
     */
    val route: String? = null,

    /**
     * platform of the frame.
     */
    val platform: String? = null,

    /**
     * List of arguments for the route.
     */
    val routeArguments: List<NativeRouteArgumentsModel>? = null,
)

/**
 * Represents an argument for a route in the native UI framework.
 */
data class NativeRouteArgumentsModel(
    /**
     * Name of the route argument.
     */
    val name: String? = null
)

/**
 * Enum representing the types of frames supported in the native UI framework.
 */
enum class FrameTypeModel(val type: String) {
    /** Standard frame type. */
    FRAME("FRAME"),

    /** Bottom sheet frame type. */
    BOTTOM_SHEET("BOTTOM_SHEET"),

    /** Dialog frame type. */
    DIALOG("DIALOG");

    companion object {
        /**
         * Converts a string representation to a corresponding enum value.
         * @param type The string representation of the frame type.
         * @return The matching enum value or FRAME if no match is found.
         */
        fun fromString(type: String) = entries.find { it.type == type } ?: FRAME
    }
}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/provider/model/NativeScaffoldModel.swift
import Foundation

/// Represents the scaffold model for Nativeblocks, containing metadata about frames and their routes.
public struct NativeScaffoldModel {
    /// A collection of frame routes defined within the scaffold.
    public let frames: [NativeFrameRouteModel]
}

/// Represents a frame route model, which defines a specific UI element and its associated metadata.
public struct NativeFrameRouteModel {
    /// The unique identifier for the frame route.
    public let id: String?

    /// The name of the frame route, used for display or identification.
    public let name: String?

    /// The type of the frame, such as a standard frame, bottom sheet, or dialog.
    public let type: FrameTypeModel?

    /// The route path for this frame, used for navigation.
    public let route: String?

    /// platform of the frame.
    public let platform: String?

    /// A collection of arguments that can be passed to this route.
    public let routeArguments: [NativeRouteArgumentsModel]?
}

/// Represents an argument for a specific route, defined by its name.
public struct NativeRouteArgumentsModel {
    /// The name of the argument.
    public let name: String?
}

/// Defines the types of frames available in Nativeblocks.
public enum FrameTypeModel: String {
    /// A standard frame type, typically a full-screen or main view.
    case frame = "FRAME"

    /// A bottom sheet type, often used for partial overlays on the screen.
    case bottomSheet = "BOTTOM_SHEET"

    /// A dialog type, used for modal interactions.
    case dialog = "DIALOG"

    /// Converts a string into a `FrameTypeModel`. Defaults to `.frame` if the string does not match a known type.
    /// - Parameter type: The string representation of the frame type.
    /// - Returns: A corresponding `FrameTypeModel` value.
    static func fromString(_ type: String) -> FrameTypeModel {
        return FrameTypeModel(rawValue: type) ?? .frame
    }
}
